File: src/stack.cpp
#include "stack.h"

#include <cstring> // Для strcmp
#include <fstream> // Для работы с файлами
#include <iostream>

using namespace std;

// Инициализация стека. Устанавливает указатель вершины стека в nullptr.
void Stack::init() { 
    top = nullptr; 
}

// Добавление элемента в стек.
void Stack::push(const string& value) {
    // Создаем новый узел стека.
    StackNode* newNode = new StackNode{value, nullptr};
    // Если стек пуст, новый узел становится вершиной стека.
    if (top == nullptr) {
        top = newNode;
    } else {
        // Иначе, находим последний элемент стека.
        StackNode* temp = top;
        while (temp->next != nullptr) {
            temp = temp->next;
        }
        // Новый узел становится следующим после последнего элемента.
        temp->next = newNode;
    }
}

// Удаление элемента из стека.
void Stack::pop() {
    // Если стек пуст, ничего не делаем.
    if (top == nullptr) {
        return;
    }
    // Если в стеке только один элемент, удаляем его.
    if (top->next == nullptr) {
        delete top;
        top = nullptr;
    } else {
        // Иначе, находим предпоследний элемент стека.
        StackNode* temp = top;
        while (temp->next->next != nullptr) {
            temp = temp->next;
        }
        // Удаляем последний элемент.
        delete temp->next;
        temp->next = nullptr;
    }
}

// Вывод содержимого стека в консоль.
void Stack::print() {
    StackNode* temp = top;
    while (temp != nullptr) {
        cout << temp->data << " ";
        temp = temp->next;
    }
    cout << endl;
}

// Освобождение памяти, занимаемой стеком.
void Stack::destroy() {
    while (top != nullptr) {
        pop(); // Последовательно удаляем элементы из стека.
    }
}

// Загрузка стека из файла.
void Stack::loadFromFile(const string& fileName) {
    ifstream file(fileName);
    string value;
    // Читаем строки из файла и добавляем их в стек.
    while (file >> value) {
        push(value);
    }
    file.close();
}

// Сохранение стека в файл.
void Stack::saveToFile(const string& fileName) {
    ofstream file(fileName);
    StackNode* temp = top;
    while (temp != nullptr) {
        file << temp->data << endl;
        temp = temp->next;
    }
    file.close();
}


// Функция для управления стеком через командную строку.
void runStack(int argc, char* argv[]) {
    Stack stack;
    stack.init();

    string fileName;
    string query;

    // Разбор аргументов командной строки.
    for (int i = 1; i < argc; i++) {
        // Поиск аргумента --file и имени файла.
        if (strcmp(argv[i], "--file") == 0 && i + 1 < argc) {
            fileName = argv[i + 1];
            i++; // Пропускаем следующий аргумент (имя файла).
        // Поиск аргумента --query и запроса.
        } else if (strcmp(argv[i], "--query") == 0 && i + 1 < argc) {
            query = argv[i + 1];
            i++; // Пропускаем следующий аргумент (запрос).
        }
    }

    stack.loadFromFile(fileName);

    string command;
    // Разделение запроса на команду и значение.
    size_t pos = query.find(' ');
    if (pos != string::npos) {
        command = query.substr(0, pos);
        query = query.substr(pos + 1);
    } else {
        command = query;
    }

    // Выполнение команды.
    if (command == "SPUSH") {
        stack.push(query);
        stack.saveToFile(fileName);
    } else if (command == "SPOP") {
        stack.pop();
        stack.saveToFile(fileName);
    } else if (command == "SPRINT") {
        stack.print();
    }

    stack.destroy();
}

File: Go/src/stack.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// StackNode — узел стека (сохраним поведение C++: push добавляет в конец списка, pop удаляет с конца)
type StackNode struct {
	data string
	next *StackNode
}

// Stack — стек, реализованный как однонаправленный список с указателем на head ("top")
type Stack struct {
	top *StackNode
}

func (s *Stack) Init() {
	s.top = nil
}

// push добавляет элемент в конец списка (так же как C++ реализация)
func (s *Stack) Push(value string) {
	if s.top == nil {
		s.top = &StackNode{data: value, next: nil}
		return
	}
	cur := s.top
	for cur.next != nil {
		cur = cur.next
	}
	cur.next = &StackNode{data: value, next: nil}
}

// pop удаляет последний элемент
func (s *Stack) Pop() {
	if s.top == nil {
		return
	}
	if s.top.next == nil {
		s.top = nil
		return
	}
	cur := s.top
	for cur.next.next != nil {
		cur = cur.next
	}
	cur.next = nil
}

func (s *Stack) Print() {
	first := true
	for cur := s.top; cur != nil; cur = cur.next {
		if !first {
			fmt.Print(" ")
		}
		first = false
		fmt.Print(cur.data)
	}
	fmt.Println()
}

func (s *Stack) Destroy() {
	s.top = nil
}

func (s *Stack) LoadFromFile(fileName string) error {
	f, err := os.Open(fileName)
	if err != nil {
		return err
	}
	defer f.Close()
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		s.Push(scanner.Text())
	}
	return scanner.Err()
}

func (s *Stack) SaveToFile(fileName string) error {
	f, err := os.Create(fileName)
	if err != nil {
		return err
	}
	defer f.Close()
	w := bufio.NewWriter(f)
	for cur := s.top; cur != nil; cur = cur.next {
		_, _ = w.WriteString(cur.data + "\n")
	}
	return w.Flush()
}

func runStack(fileName string, fullQuery string) {
	st := Stack{}
	st.Init()
	_ = st.LoadFromFile(fileName)

	parts := strings.SplitN(fullQuery, " ", 2)
	command := parts[0]
	arg := ""
	if len(parts) > 1 {
		arg = parts[1]
	}

	switch command {
	case "SPUSH":
		st.Push(arg)
		_ = st.SaveToFile(fileName)
	case "SPOP":
		st.Pop()
		_ = st.SaveToFile(fileName)
	case "SPRINT":
		st.Print()
	default:
		fmt.Fprintln(os.Stderr, "Неизвестная команда:", command)
	}

	st.Destroy()
}
